src/sinks/datadog/mod.rs

```diff
@@ -46,3 +47,47 @@ async fn healthcheck(
         other => Err(HealthcheckError::UnexpectedStatus { status: other }.into()),
     }
 }
+
+#[derive(Debug, Snafu)]
+pub enum DatadogApiError {
+    #[snafu(display("Server responded with an error."))]
+    ServerError,
+    #[snafu(display("Failed to make HTTP(S) request: {}", error))]
+    HttpError { error: HttpError },
+    #[snafu(display("Client sent a payload that is too large."))]
+    PayloadTooLarge,
+    #[snafu(display("Client request was not valid for unknown reasons."))]
+    BadRequest,
+    #[snafu(display("Client request was forbidden."))]
+    Forbidden,
+}
+
+pub const fn is_retriable_error(error: &DatadogApiError) -> bool {
+    match *error {
+        DatadogApiError::HttpError {
+            error: HttpError::BuildRequest { .. },
+        }
+        | DatadogApiError::HttpError {
+            error: HttpError::MakeProxyConnector { .. },
+        }
+        | DatadogApiError::BadRequest
+        | DatadogApiError::PayloadTooLarge => false,
+        // This retry logic will be expanded further, but specifically retrying unauthorized
+        // requests and lower level HttpErrorsfor now.
+        // I verified using `curl` that `403` is the respose code for this.
+        //
+        // https://github.com/vectordotdev/vector/issues/10870
+        // https://github.com/vectordotdev/vector/issues/12220
+        DatadogApiError::HttpError {
+            error: HttpError::CallRequest { .. },
+        }
+        | DatadogApiError::HttpError {
+            error: HttpError::BuildTlsConnector { .. },
+        }
+        | DatadogApiError::HttpError {
+            error: HttpError::MakeHttpsConnector { .. },
```

Comment:
It strikes me that we may want a way to have `HttpError` report whether it is retriable or not, and then we proxy to that.

It also strikes me that we might want this as an inherent method on `DatadogApiError` itself rather than a free function, since doing the above for `HttpError` would start to get confusing from a `use ...::http::is_retriable_error, ...::datadog::is_retriable_error` standpoint.